ngcc: cleanup outdated artifacts

petebacondarwin opened the pull request:
Fixes #35082

I looked at various approaches of opting it or doing a special one-off cleaning pass, but I found that it  we can completely automate the cleaning so that the developer doesn't even need to be aware of it - and by always doing it in this way we are avoiding a whole range of awkward corner cases where we need to decide whether (and when) to run the clean up.

Note that this approach of using the backup files as the markers for cleanup means that we must clean whole packages. We cannot simply clean a single entry-point to a package. I think this is actually reasonable. Entry-points that have not been processed with be left untouched anyway by the cleaning. And if you have one entry-point that is outdated then you would want to fix all other outdated ones in the package anyway.

One thing this doesn't do is guarantee that, after you have run ngcc, that all packages in the `node_modules` folder are clean. Only those that would have been processed by ngcc are cleaned. But since we will now always clean on every run of ngcc, if there are any packages that were not cleaned on the one run of ngcc (because they were not needed) they will just get cleaned up they are needed. (The use case here is CLI).

One thing that we might need to check with the CLI team is whether they need to track changes to files due to the clean up. I think they track files that are changed by ngcc in general already so perhaps this will just work out of the box.

alxhub commented on packages/compiler-cli/ngcc/src/writing/cleaning/cleaning_strategies.ts:
```diff
@@ -0,0 +1,62 @@
+  clean(path: AbsoluteFsPath, _basename: PathSegment): void {
+    const packageJson = JSON.parse(this.fs.readFile(path));
+    cleanPackageJson(packageJson);
+    this.fs.writeFile(path, JSON.stringify(packageJson));
```
Do we do this already? `JSON.stringify` will return a one-line blob with no whitespace. It's not a good experience for users if ngcc compresses their package.json like that.

I think we should at _least_ do 
```suggestion
    this.fs.writeFile(path, JSON.stringify(packageJson, null, 2));
```
which will apply some prettier formatting. But I'd prefer if we could avoid reformatting any parts of the file besides the keys we touch.

petebacondarwin replied:
Good spot. Currently we do this:

```
      this.fs.writeFile(path, `${JSON.stringify(packageJson, null, 2)}\n`);
```

I will update this bit of code to follow this pattern - but also not to write the file at all if we didn't change it.

gkalpak replied:
That's what the `PackageJsonUpdater` does (`JSON.stringify(..., null, 2)` + `'\n'`). I think npm/yarn may do the same, so I believe it is fine in this case.

gkalpak commented on packages/compiler-cli/ngcc/src/writing/cleaning/utils.ts:
```diff
@@ -0,0 +1,23 @@
+*
+* @param fs the current filesystem
+* @param path the path to check
+*/
```
Incorrect indentation.

gkalpak commented on packages/compiler-cli/ngcc/src/writing/cleaning/utils.ts:
```diff
@@ -0,0 +1,23 @@
+export function isLocalDirectory(fs: FileSystem, path: AbsoluteFsPath): boolean {
+  if (fs.exists(path)) {
+    const stat = fs.lstat(path);
+    return (stat.isDirectory() && !stat.isSymbolicLink());
```
The `isSymbolicLink()` check shouldn't be needed, since we are using `lstat`.

gkalpak commented on packages/compiler-cli/ngcc/src/writing/cleaning/cleaning_strategies.ts:
```diff
@@ -0,0 +1,62 @@
+  clean(path: AbsoluteFsPath, _basename: PathSegment): void {
+    const packageJson = JSON.parse(this.fs.readFile(path));
+    cleanPackageJson(packageJson);
+    this.fs.writeFile(path, JSON.stringify(packageJson));
```
FWIW, I believe it would be preferable to do the `package,json` modifications through `PackageJsonUpdater`, but this should not block this PR anyway.

petebacondarwin replied:
I guess the benefit is that it will be more consistent in writing the files out...

petebacondarwin replied:
But unfortunately these modification require removal of properties from package.json, which the current updater doesn't handle.

So I think this is too big an extra for this PR.

gkalpak commented on packages/compiler-cli/ngcc/src/writing/cleaning/package_cleaner.ts:
```diff
@@ -0,0 +1,76 @@
+        }
+      }
+      // Recurse into subdirectories (note that a cleaner may have removed this path)
+      if (basename !== 'node_modules' && isLocalDirectory(this.fs, path)) {
```
Nit: We could probably move the `basename !== 'node_modules'` check at the top of the `for` loop and save the work associated with building the full path and iterating over all cleaners for this path.

petebacondarwin replied:
That makes the assumption that no cleaning strategies will want to do something with `node_modules` :-P

(But since that is not the case, I can move it).

gkalpak reviewed (approved):
Nice clean-up :tada:
I left several comments, but almost all of them are non-blocking (except for  the ones about adding `fs-extra` as a dependency). 

<sub>I noticed that you pushed some new commits as I was about to finish my review, so some of my comments might have already been addressed.</sub>

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/build_marker_spec.ts:
```diff
@@ -194,5 +195,103 @@ runInEachFileSystem(() => {
     });
+
+    describe('needsCleaning()', () => {
+      it('should return true if any format has been compiled with a different version.', () => {
```
No idea. One probably sneaked in there and then I probably cut and pasted it everywhere.

IgorMinar reviewed (commented):
> Note that this approach of using the backup files as the markers for cleanup means that we must clean whole packages. We cannot simply clean a single entry-point to a package. I think this is actually reasonable. Entry-points that have not been processed with be left untouched anyway by the cleaning. And if you have one entry-point that is outdated then you would want to fix all other outdated ones in the package anyway.

sgtm, package-level consistency is desirable

> One thing this doesn't do is guarantee that, after you have run ngcc, that all packages in the node_modules folder are clean. Only those that would have been processed by ngcc are cleaned. But since we will now always clean on every run of ngcc, if there are any packages that were not cleaned on the one run of ngcc (because they were not needed) they will just get cleaned up they are needed. (The use case here is CLI).

that is ok, we don't need node_module level consistency. if someone cares about that, they can use yarn instead of npm.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/build_marker.ts:
```diff
@@ -6,11 +6,50 @@
+ */
+export function needsCleaning(packageJson: EntryPointPackageJson): boolean {
+  return Object.keys(packageJson.__processed_by_ivy_ngcc__ || {})
+      .some(property => packageJson.__processed_by_ivy_ngcc__ ![property] !== NGCC_VERSION);
```
Nice!

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/build_marker.ts:
```diff
@@ -6,11 +6,50 @@
+}
+
+/**
+ * Clean an build marker artifacts from the given `packageJson` object.
```
Actually I was going for:

```suggestion
 * Clean any build marker artifacts from the given `packageJson` object.
```

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/BUILD.bazel:
```diff
@@ -9,7 +9,9 @@ ts_library(
         "//packages:types",
+        "@npm//@types/fs-extra",
         "@npm//@types/node",
+        "@npm//fs-extra",
```
It is - I added in one of the fixup commits.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/cached_file_system.ts:
```diff
@@ -88,6 +88,12 @@ export class CachedFileSystem implements FileSystem {
 
+  removeDir(path: AbsoluteFsPath): void {
+    this.delegate.removeDir(path);
+    this.existsCache.set(path, false);
```
Hmm good point...

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system.ts:
```diff
@@ -131,6 +131,21 @@ export abstract class MockFileSystem implements FileSystem {
+    }
+    if (!isFolder(entity[basename])) {
+      throw new MockFileSystemError(
+          'ENOTDIR', path, `Unable to remove "${path}" - it is not a directory.`);
```
Hmm... I was copying the standard NodeJS `rmDirSync()` behaviour. But since the recursive form of that is only available in Node 12+ then we couldn't use it here. The fs-extra version can indeed remove files too.

How about we change the function name to `removeDeep()` and allow it to remove files or directories?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/test/BUILD.bazel:
```diff
@@ -12,6 +12,7 @@ ts_library(
         "//packages:types",
         "//packages/compiler-cli/src/ngtsc/file_system",
         "//packages/compiler-cli/src/ngtsc/file_system/testing",
+        "@npm//@types/fs-extra",
```
The vagaries of Bazel deps always confuse me.

So to "build" this TS library we only need the typings.
But to "run" code that uses the built JS of this library you need the actual package.
So I think that for this rule, it is correct. But then for things like the unit tests we need to add in the actual package.

petebacondarwin replied:
Actually, what I mean is that the runtime package gets used because the src BUILD.bazel mentions it.
We just need the typings here so that Bazel doesn't complain that we have a non-direct dependency.

filipesilva reviewed (approved):
I don't think the CLI needs followup changes from this PR. From the CLI point of view, if ngcc runs we just purge the package.json from our file systems and reload it.

petebacondarwin committed:
fixup! refactor(ngcc): add additional build marker helpers

petebacondarwin committed:
refactor(ngcc): add additional build marker helpers

petebacondarwin committed:
refactor(ngcc): do not crash if package build version is outdated

Now `hasBeenProcessed()` will no longer throw if there
is an entry-point that has been built with an outdated
version of ngcc.

Instead it just returns `false`, which will include it in this
processing run.

This is a precursor to adding functionality that will
automatically revert outdate build artifacts.

petebacondarwin committed:
refactor(ngcc): remove unused code

petebacondarwin committed:
fixup! refactor(ngcc): add additional build marker helpers

petebacondarwin committed:
fixup! feat(ngcc): automatically clean outdated ngcc artifacts

petebacondarwin committed:
fixup! refactor(compiler-cli): add `removeDir()` to `FileSystem`

petebacondarwin committed:
feat(ngcc): automatically clean outdated ngcc artifacts

If ngcc gets updated to a new version then the artifacts
left in packages that were processed by the previous
version are possibly invalid.

Previously we just errored if we found packages that
had already been processed by an outdated version.

Now we automatically clean the packages that have
outdated artifacts so that they can be reprocessed
correctly with the current ngcc version.

Fixes #35082

petebacondarwin committed:
fixup! feat(ngcc): automatically clean outdated ngcc artifacts

petebacondarwin committed:
fixup! refactor(compiler-cli): add `removeDir()` to `FileSystem`

petebacondarwin committed:
fixup! refactor(compiler-cli): add `removeDir()` to `FileSystem`

petebacondarwin committed:
refactor(ngcc): export magic strings as constants

These strings will be used when cleaning up outdated
packages.

petebacondarwin committed:
refactor(compiler-cli): add `removeDir()` to `FileSystem`

petebacondarwin committed:
fixup! refactor(compiler-cli): add `removeDir()` to `FileSystem`

petebacondarwin committed:
fixup! refactor(ngcc): add additional build marker helpers

petebacondarwin committed:
build: temporarily disable components-repo-unit-tests

Disabling this job until the fix at
https://github.com/angular/components/pull/18355 or similar lands.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -727,7 +727,7 @@ jobs:
       - run:
           name: "Running `angular/components` unit tests"
-          command: ./scripts/ci/run_angular_components_unit_tests.sh
+          command: ./scripts/ci/run_angular_components_unit_tests.sh | exit 0
```
in a follow up PR: can you add a FIXME comment here as well so that we don't forget to remove `| exit 0` plz

petebacondarwin replied:
Oops, I meant to remove that, as I commented out the job...

petebacondarwin replied:
https://github.com/angular/angular/pull/35090/commits/285e841736451bf1bb297fa95c775b04ef69d80c

IgorMinar reviewed (approved):
at high level this LGTM, I requested one more change but let's do that in a separate PR so that we don't block this any more
